### DISCUSSION TITLE
Config file format poll

### DISCUSSION THREAD

**Discussion by MikeNeilson** [opendcs/opendcs#160] [Polls]
@opendcs/opendcs-core-devs @opendcs/opendcs-triange 

So, with the discussion of the LRGS service and improvements, and my starting to create MQTT LRGS Input source and looking at how to configure it, the combination of XML and sort of but not really java .properties files it a bit... confusing. It would be good to decide on a single format and enforce it going forward.

Going forward we should standardize on configuration file formats (not necessarily changing them in 7.x but definitely for 8).

Once we've standardized it should be easier to do things like injecting properties from other sources and such, like the environment.

Some arguments from someone else: https://octopus.com/blog/state-of-config-file-formats

My inclination is XML.  while not my favorite markup language in the work it is very descriptive and is flexible enough to allow configuration elements for the various implementations, even making it easier to pass them directly onto the implementation instead of a single parameter parser for everything and everyone.

**Comment by agilmore2** [comment]
If we're going to make a change away from the current properties file
formats, I'd prefer to have it still be straightforward for users to
comprehend, which is where XML fails.

Seems everything I see nowadays uses JSON. I don't like how comments are
kludgy in it, but otherwise I prefer it, especially over XML and YAML.

**Reply by MikeNeilson** [reply]
Fair, I like xml for it's consistency but it can obnoxious for the human to use.

I've used a few JSON based configs, and I'm not especially a fan. same with YAML.

I haven't voted yet either to get the discussion going but I was thinking java .properties files. It's the most direct (except for DRGS.conf and ddsrecv.conf) replacement as we already use "properties" files, it's just that they were started before the standard Java Properties class and don't necessarily conform to that standard.

There's also the question of exactly how much should go *in* that initial properties file vs elements in the database. For example the LRGS. 
*if* we move the lrgs configuration and archive to being in, we'll say JDBC accessible database, should there really be more than just the minimum to connect to that database and the rest of the configuration is there. 

Or may it's a combination. The archive itself should be able to be an independent database if desired.

**Reply by cappelaere** [reply]
I would argue that some people may consider not a database but a data streaming service... that will not only distribute the data but allow for the processing of the data and eventually its storage. :)

**Reply by MikeNeilson** [reply]
While it may need to change, the "LRGS" is currently a data aggregator that collects from various streamed or polled source so that downstream components (DECODES) only need to talk to one system. But that is conversation for https://github.com/opendcs/opendcs/discussions/152.

The question here is, regardless of which component of OpenDCS we're using, what the heck do we want to properly enforce as a baseline configuration file format? (as it's gone between a combination of xml and almost java properties.)

**Comment by agilmore2** [comment]
I'd be ok going .properties format as well.

**Comment by DanielTOsborne** [comment]
My primary vote is for proper Java .properties format. It's the closest to what is already used in some places.
My second vote would be for XML, as it's also a common format, and while not as human readable, is very descriptive and can be commented heavily (as can .properties).

Comments are officially excluded from the JSON spec, so I do not recommend using that.

**Comment by adamkorynta** [comment]
If Java Properties is sufficient to meet all the requirements it makes sense to use it. If added complexity is needed then something like TOML is probably not robust enough anyways.

I'll throw the lightbend-config (https://github.com/lightbend/config) library into the conversation as it has full API support for merging properties sources together, validation of properties, and immutability. It reads java properties files as well as its own HOCON format which adds comment support JSON.

**Reply by DanielTOsborne** [reply]
I definitely like the idea of merging definitions, kind of like an "include" for configs.

**Comment by mikecovesw** [comment]
The "lrgs.conf" is a Java properties file.
The two files "drgsconf.xml" and "ddsrecv.conf" are XML because an extra layer of depth was needed. They each have settings for multiple connections. 
I would have preferred everything in one file (Java properties is so easy to view and edit) but I would have needed complex names like "drgs.con1.hostname". XML seemed a better fit at the time.

**Comment by MikeNeilson** [comment]
That's understandable, though i suspect at this point people are more
acustomed to the dotted name syntax.

Well, and those who aren't probably want to just use a gui interface
anyways
